HIP/hip-358.md

```diff
@@ -0,0 +1,123 @@
+
+As a smart contract developer, I want to be able to create a native HTS fungible token with custom fees through a solidity contract call.
+
+As a smart contract developer, I want to be able to create a native HTS non-fungible token through a solidity contract call.
+
+As a smart contract developer, I want to be able to create a native HTS non-fungible token with custom fees through a solidity contract call.
+
+## Specification
+
+### EVM Precompile extension
+
+The [Solidity file for development](../assets/hip-206/solidity/IHederaTokenService.sol) is updated with new types and function signatures that the precompile will respond to. It is included in this HIP by reference.
+
+### Solidity Function Signatures 
+
+The following structs are defined to simplify the interface between solidity and the HTS precompile: 
+
+| Name                | Definition                                                                  |
+| ------------------- | ----------------------------------------------------------------------------|
+| `KeyValue`          | `(bool, address, bytes, bytes, address)`                                    |
+| `Expiry`            | `(uint, address, uint)`                                                     |
+| `TokenKey`          | `(uint, KeyValue)`                                                          |
+| `HederaToken`       | `(string, string, address, string, bool, uint32, bool, TokenKey[], Expiry)` |
+| `FixedFee`          | `(uint32, address, bool, bool, address)`                                    |
+| `FractionalFee`     | `(uint32, uint32, uint32, uint32, bool, address)`                           |
+| `RoyaltyFee`        | `(uint32, uint32, uint32, address, bool, address)`                          |
+
+Two functions are defined for creating fungible and non-fungible tokens each. One is a simple version with just the required fields and the other a full version supporting custom fees as well. The ABI signature and hashes for each call are as follows:
+
+| hash       | effective signature                                                                       | return           |
+| ---------- | ------------------------------------------------------------------------------------------|------------------|
+| `7812a04b` | `createFungibleToken(HederaToken, uint, uint)`                                            | `(int, addess)`  |
+| `4c381ae7` | `createFungibleTokenWithCustomFees(HederaToken, uint, uint, FixedFee[], FractionalFee[])` | `(int, addess)`  |
+| `9dc711e0` | `createNonFungibleToken(HederaToken)`                                                     | `(int, addess)`  |
+| `5bc7c0e6` | `createNonFungibleTokenWithCustomFees(HederaToken, FixedFee[], RoyaltyFee[])`             | `(int, addess)`  |
+
+### Precompile Gas Costs
+
+In order to allow creation of token through the precompiled contract, a new mechanism for payment is introduced - precompile calls are partially payed by Hbar value transfers.
```

Comment:
We will have to update `com.hedera.services.store.contracts.HederaWorldState#TOKEN_CALL_REDIRECT_CONTRACT_BINARY` - make it payable.  The current evm code will fail when any value is passed in.  We will need to move the value check to another part of the HTS precompile.